### ISSUE TITLE
Crash: Multiple instances of same OnlineRepository, set same requirements causes race conditions. 

### ISSUE THREAD

**Issue by levibostian** [levibostian/Teller-iOS#32]
Teller version: commit [51bab3a8527824fefea2625f2383a534cd08de5a](https://github.com/levibostian/Teller-iOS/commit/51bab3a8527824fefea2625f2383a534cd08de5a)

Teller causes a crash in this section of code below:

```swift
let hasEverFetchedDataBefore = self.syncStateManager.hasEverFetchedData(tag: requirements.tag) // Get value before calling `updateAgeOfData`.
self.syncStateManager.updateAgeOfData(tag: requirements.tag)

if !hasEverFetchedDataBefore {
    self.currentStateOfData.changeState({ try! $0.successfulFirstFetch(timeFetched:     self.syncStateManager.lastTimeFetchedData(tag: requirements.tag)!) })
} else {
    self.currentStateOfData.changeState({ try! $0.successfulFetchingFreshCache(timeFetched: Date()) })
}
```

Specifically the line: `self.currentStateOfData.changeState({ try! $0.successfulFetchingFreshCache(timeFetched: Date()) })`.

What is the crash? It is because `hasEverFetchedDataBefore` is set to *true* but the current state of data is *first time fetching*. Having these 2 values out of sync is bad news. We have a data sync issue there. 

This is how you can reproduce this bug: 

1. Create 2+ instances of the same `OnlineRepository` subclass. 
2. Set the requirements of both instances to the same tag. 
3. That's it. 2 fetch calls will be started (1 for each instance. Teller only guarantees 1 fetch call per instance.) and each fetch call result comes in. 
4. The crash will happen if 1 fetch call comes in and writes to the `SyncStateManager` before the other fetch call is complete.  

---

### Temp fix: 

Only use `SyncStateManager` when `requirements` set on an `OnlineRepository`. That's the purpose of the `SyncStateManager` anyways: to persist that state of data beyond memory by writing to the `UserDefaults`. After we refer to `SyncStateManager` once to get the state of data, we do not need it. 

### Better fix: 

First, perform the *Temp fix*. Then, we could share the refresh call between *all* of the `OnlineRepository` instances for the same tag. So that way when the 2+ instances are created and the requirements are set, they all share the same fetch call. I do not believe it's a good idea to share any other resource together, but sharing the fetch call would be a benefit to only make sure only 1 fetch call is called. 

We need to still guarantee that the nodes being traveled in the `OnlineData` state machine are valid. So a solution such as "after a successful fetch, check the `SyncStateManager.hasEverFetchedData()` and just skip the `first fetch` if the data has already been fetched from another instance". That would skip nodes in the state machine and then the UI could be out of sync. We guarantee to the user of Teller that the nodes are traveled in a valid order. 

This "better fix" also means that we would have multiple calls to `saveData` which is not a huge deal, but it needs to probably be thread safe to avoid multiple saves at 1 time (even though a DB should be made to prevent this as an issue). 

---

This also makes me think, perhaps we could do a refactor to help prevent bugs in the future. The SyncStateManager and the `currentStateOfData` need to be: (1) always reflecting the same state together and (2) thread-safe. 

So, maybe we could combine the two together somehow? Right now, they are not coupled and are independently updated. But when you think about it, they *both* are in charge of determining the state of the cache data. SyncStateManager is only required when requirements are set then after that, we can use the `currentStateOfData` (as long as it's thread-safe) to get the state of data.

**Comment by levibostian** [comment]
Commit: https://github.com/levibostian/Teller-iOS/commit/298ca54e983583f27d70cbf6cc6dd43071cc6843 is an attempt at making `currentStateOfData` thread-safe.

**Comment by levibostian** [comment]
PR: https://github.com/levibostian/Teller-iOS/pull/33 Fixes this bug using the "Temp fix:" plan.

**Comment by levibostian** [comment]
I will be closing this issue as the bug has been resolved with commit `0a149569bc3a40a31047cb178f77c30cd6fd0cd3`. 

I will refer to these 2 new issues that break down the "better fix" that can be worked on in the future but are not time sensitive. 

1. https://github.com/levibostian/Teller-iOS/issues/35
2. https://github.com/levibostian/Teller-iOS/issues/34
